### PR TITLE
Paths keyword

### PR DESCRIPTION
**Pull request by psvkaushik**

### CODE REVIEW

**Comment by psvkaushik** [comment]
In the commit with commit message "version 1.0.8, the file with type declarations for each element of the PATHS keyword wsa added.

**Review by balhoff — CHANGES_REQUESTED** [review]
@psvkaushik can you remove the changes to `package.json` and `package-lock.json`? I'd like to try a different method for depending on local modifications in the other projects.

Also, please revert the changed filenames for `a*.sparql`, `a?.sparql`, `a*.json`, and `a?.json`.

**Comment by psvkaushik** [comment]
I've put the package.json and package-lock.json in gitignore in this commit.
I cannot change the file names due to the OS issue with windows, could not clone the repo due to the same issues.

**Comment by balhoff** [comment]
> I've put the package.json and package-lock.json in gitignore in this commit. I cannot change the file names due to the OS issue with windows, could not clone the repo due to the same issues.

When making git commits, try to add only the specific changes you intend. So even if Windows messes up the file name, you don't need to commit that change. I usually review everything that has changed and only add the particular files or pieces of files that I want for a particular commit (in the event that some other things inadvertently changed).

I've changed back the names of the `?` and `*` files through the GitHub UI. Adding the package files to gitignore doesn't affect the changes already committed here, so I would revert those to their previous state. I think it would be best not to alter gitignore either, since this project in general does want those files committed. You will just need to ignore local changes to them when you make git commits.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,322 @@
+// Type definitions for sparqljs 3.1
```
Comment (balhoff) [inline index.d.ts:1]: Just realized this file is added. Is it automatically generated? If so we probably shouldn't commit.

File: lib/sparql.jison [hunk 2]
```diff
@@ -728,7 +736,79 @@ QueryOrUpdate
       return $2;
     }
     ;
+PathsQuery
+    : 'PATHS' PathProperties LimitOffsetClauses?
+    {
+      // Construct the query object for PATHS
+      let pathsQuery = {
```
Comment (balhoff) [inline lib/sparql.jison]: Can this be a const?

File: lib/sparql.jison [hunk 3]
```diff
@@ -728,7 +736,79 @@ QueryOrUpdate
...
+    | ListOfIris
+    
+    ;
+
+ListOfIris
+    : '[' iri (',' iri)* ']'
+    {
+      $$ = [$2].concat($3.map(pair => pair[1]));
+    }
+    ;
```
Comment (balhoff) [inline lib/sparql.jison]: Are these used for anything?

File: main.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// main.js
```
Comment (balhoff) [inline main.js]: Don't commit this file.

**Review by balhoff — CHANGES_REQUESTED** [review]
@psvkaushik I noted a few more changes to make. Rather than committing the main file, add some tests so that we can verify functionality. Looking good overall!

File: index.d.ts (hunk 1 above)
Comment (psvkaushik) [reply index.d.ts:1]: It is not automatically generated, but without this in the package, the other repos are throwing errors. In sparqlalgebrajs and comunica they're asking for type definitions.
So is it ok if I commit it from the next iteration?

**Comment by psvkaushik** [comment]
These changes are based on the review comments, I'll push the changes consisting of stardog like grammar in the next commits.

File: index.d.ts (hunk 1 above)
Comment (balhoff) [reply index.d.ts:1]: How was it generated?

**Review by balhoff — CHANGES_REQUESTED** [review]
Some cleanup stuff that needs to be fixed before we can merge:

- It looks like `a*.sparql`, `a?.sparql`, `a*.json`, and `a?.json` are all deleted from the repo now.  These should not be altered.
- Don't modify `package.json` or `package-lock.json` (the commits from #3 should be merged into this branch also)
- Don't commit `main.js`
- Don't modify `.gitignore`
- Not sure about `index.d.ts`. If it wasn't already committed, it doesn't seem correct to add it as part of this pull request. Adding it is a different topic.

**Comment by psvkaushik** [comment]
I think changing gitignore removed the files. Hopefully the file names are not changed. Reverted the changes made to package.json and package-lock.json. From now on will only add the sparql.jison file to which I am making the changes.

File: lib/sparql.jison [hunk 5]
```diff
@@ -676,7 +685,7 @@ SPACES_COMMENTS       (\s+|{COMMENT}\n\r?)+
 %%
 
 QueryOrUpdate
-    : Prologue ( Query /* [1] QueryUnit */ | Update? | Path ) EOF
+    : Prologue ( Query /* [1] QueryUnit */ | Update? | Path | PathsQuery) EOF
```
Comment (balhoff) [inline lib/sparql.jison:688]: @psvkaushik move this to be a peer to SELECT, CONSTRUCT, DESCRIBE, and ASK in `Qry`.

File: lib/sparql.jison [hunk 6]
```diff
@@ -728,6 +737,123 @@ QueryOrUpdate
...
+      $$ = { type: 'iri', value:  { termType: 'NamedNode', value: $2 } };
+    }
+    | '=' WhereClause
+    {
+      $$ = { type: 'pattern', value: $2.where };
+    }
+ 
+    ;
+PathViaValue
+    : iri
```
Comment (balhoff) [inline lib/sparql.jison:843]: This should be a property path rather than just IRI.

File: lib/sparql.jison [hunk 7]
```diff
@@ -728,6 +737,123 @@ QueryOrUpdate
...
+    | 'END'  Var PathValue?
+    {
+        $$ = { property: 'end',
+        value: {
+          var: { termType: 'Variable', value: $2.value },
+          value: { type: $3.type, value: $3.value } 
+      } 
+        };
+    }
+    | 'VIA' (Var|PathViaValue)
```
Comment (balhoff) [inline lib/sparql.jison:793]: Why not make the Var one of the possibilities in PathViaValue?

File: lib/sparql.jison [hunk 8]
```diff
@@ -728,6 +737,123 @@ QueryOrUpdate
...
+    {
+      $$ = { type: 'iri', value:{ termType: 'NamedNode', value: $1 }};
+    }
+    | WhereClause
+    {
+      $$ = { type: 'pattern', value: $1.where };
+    }
+    |
+    {
+      $$ = null;
```
Comment (balhoff) [inline lib/sparql.jison:853]: What is this for?

**Review by balhoff — CHANGES_REQUESTED** [review]
Looking much better! Nice work on adding all the tests. I put a few more comments. 😅

**Comment by balhoff** [comment]
Closing this PR in favor of keeping a separate feature branch going.

**Pull request closed without merging**